// react-apollo/lib/react-apollo.cjs.js
exports.__esModule = true;
exports.renderToStringWithData = exports.getMarkupFromTree = exports.getDataFromTree = exports.useSubscription = exports.useQuery = exports.useMutation = exports.useLazyQuery = exports.useApolloClient = exports.withSubscription = exports.withQuery = exports.withMutation = exports.withApollo = exports.graphql = exports.Subscription = exports.Query = exports.Mutation = exports.resetApolloContext = exports.getApolloContext = exports.ApolloProvider = exports.ApolloConsumer = void 0;

var _reactCommon = require("@apollo/react-common");

exports.ApolloConsumer = _reactCommon.ApolloConsumer;
exports.ApolloProvider = _reactCommon.ApolloProvider;
exports.getApolloContext = _reactCommon.getApolloContext;
exports.resetApolloContext = _reactCommon.resetApolloContext;

var _reactComponents = require("@apollo/react-components");

exports.Mutation = _reactComponents.Mutation;
exports.Query = _reactComponents.Query;
exports.Subscription = _reactComponents.Subscription;

var _reactHoc = require("@apollo/react-hoc");

exports.graphql = _reactHoc.graphql;
exports.withApollo = _reactHoc.withApollo;
exports.withMutation = _reactHoc.withMutation;
exports.withQuery = _reactHoc.withQuery;
exports.withSubscription = _reactHoc.withSubscription;

var _reactHooks = require("@apollo/react-hooks");

exports.useApolloClient = _reactHooks.useApolloClient;
exports.useLazyQuery = _reactHooks.useLazyQuery;
exports.useMutation = _reactHooks.useMutation;
exports.useQuery = _reactHooks.useQuery;
exports.useSubscription = _reactHooks.useSubscription;

var _reactSsr = require("@apollo/react-ssr");

exports.getDataFromTree = _reactSsr.getDataFromTree;
exports.getMarkupFromTree = _reactSsr.getMarkupFromTree;
exports.renderToStringWithData = _reactSsr.renderToStringWithData;
